File: LeetCodeCPlusPlus/py81.cpp
#include <vector>
using namespace std;
class Solution {
public:
    //C++最简洁的二分法分类讨论
//每次二分，左半部分和右半部分至少有一边是有序的，以此为条件可以分成两种情况：
//1、左半边是有序的
//(1) target落在左半边
//(2) otherwise
//2、右半边是有序的
//(1) target落在右半边
//(2) otherwise
//综上所述，一共两种可能性，这两种情况各自又有两种可能性，代码如下：
bool search(vector<int>& nums, int target) {
        int l = 0, r = nums.size()-1;
        while(l<=r){
            //处理重复数字
            while(l<r&&nums[l]==nums[l+1]) ++l;
            while(l<r&&nums[r]==nums[r-1]) --r;
            int mid = l+(r-l)/2;
            if(nums[mid]==target) return true;
            //左半部分有序
            if(nums[mid]>=nums[l]){
                if(target<nums[mid]&&target>=nums[l]) r = mid-1;//target落在左半边
                else l = mid+1;
            }
            else{//右半部分有序
                if(target>nums[mid]&&target<=nums[r]) l = mid+1;//target落在右半边
                else r = mid-1;
            }
        }
        return false;
    }
};

File: LeetCodePython/py81.py
# 搜索旋转链表, 因为可以包含相同数值，因此没办法确定哪一部分有序，因此在前面增加一种判断，nums[mid] == nums[left] == nums[right]
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        
        left = 0
        right = len(nums) - 1
        while left <= right:
            #print(left, right)
            mid = left + (right - left) // 2
            # 等于目标值
            if nums[mid] == target:return True
            
            if nums[mid] == nums[left] == nums[right]:
                left += 1
                right -= 1
            # 在前部分
            elif nums[left] <= nums[mid]:
                if nums[left] <= target < nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                if nums[mid] < target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        return False

# 作者：powcai
# 链接：https://leetcode-cn.com/problems/two-sum/solution/er-fen-by-powcai/
# 来源：力扣（LeetCode）
# 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。